File: tests/cpp_test/n2_test.cpp
// Copyright 2017 Kakao Corp. <http://www.kakaocorp.com>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <vector>
#include "gtest/gtest.h"

#include "n2/hnsw.h"
#include "n2/distance.h"
#include "n2/min_heap.h"

class CppApiTest : public::testing::Test {
    protected:
        virtual void SetUp() {}
        virtual void TearDown() {}
};

TEST_F(CppApiTest, SearchByVectorTest) {
    n2::Hnsw index(3, "angular");
    index.AddData(std::vector<float>{0, 0, 1});
    index.AddData(std::vector<float>{0, 1, 0});
    index.AddData(std::vector<float>{0, 0, 1});
    index.Build(5, 10);
    std::vector<std::pair<int, float> > result;
    int ef_search = 3*10;
    index.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);    
    EXPECT_EQ(3, result.size());
    n2::Hnsw index2(3, "angular");
    index2.AddData(std::vector<float>{0, 0, 1});
    index2.AddData(std::vector<float>{0, 1, 0});
    index2.AddData(std::vector<float>{0, 0, 1});
    index2.Build(5, 10);
    std::vector<std::pair<int, float> > result2;
    index2.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result2);
    EXPECT_EQ(3, result2.size());    
}

TEST_F(CppApiTest, SearchByIdTest) {
    n2::Hnsw index(3);
    index.AddData(std::vector<float>{2, 1, 0});
    index.AddData(std::vector<float>{1, 2, 0});
    index.AddData(std::vector<float>{0, 0, 1});
    index.Build(5, 10, 150, 1, 0, n2::NeighborSelectingPolicy::HEURISTIC, n2::GraphPostProcessing::SKIP);
    std::vector<std::pair<int, float> > result;
    int ef_search = 3*10;
    index.SearchById(0, 3, ef_search, result);
    EXPECT_EQ(3, result.size());
    EXPECT_EQ(0, result[0].first);
    EXPECT_EQ(1, result[1].first);
    EXPECT_EQ(2, result[2].first);
    result.clear();
    index.SearchById(1, 3, ef_search, result);
    EXPECT_EQ(3, result.size());
    EXPECT_EQ(0, result[1].first);
    EXPECT_EQ(1, result[0].first);
    EXPECT_EQ(2, result[2].first);
    result.clear();
}

TEST_F(CppApiTest, CopyOperatorTest) {
    n2::Hnsw* origin = new n2::Hnsw(3, "angular");
    origin->AddData(std::vector<float>{0, 0, 1});
    origin->AddData(std::vector<float>{0, 1, 0});
    origin->AddData(std::vector<float>{0, 0, 1});
    origin->Fit();
    std::vector<std::pair<int, float> > result;
    int ef_search = 3*10;
    origin->SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);
    EXPECT_EQ(3, result.size());
    n2::Hnsw* copy = new n2::Hnsw(*origin);
    delete origin;
    std::vector<std::pair<int, float> > result2;
    copy->SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result2);
    EXPECT_EQ(3, result2.size());
    delete copy;
}

TEST_F(CppApiTest, AssignOperatorTest) {
    n2::Hnsw origin(3, "angular");
    origin.AddData(std::vector<float>{0, 0, 1});
    origin.AddData(std::vector<float>{0, 1, 0});
    origin.AddData(std::vector<float>{0, 0, 1});
    origin.Fit();
    std::vector<std::pair<int, float> > result;
    int ef_search = 3*10;
    origin.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);
    EXPECT_EQ(3, result.size());
    n2::Hnsw copy(3, "angular");
    copy = origin;
    std::vector<std::pair<int, float> > result2;
    copy.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result2);
    EXPECT_EQ(3, result2.size());
}

TEST_F(CppApiTest, MoveAssignOperatorTest) {
    n2::Hnsw origin(3, "angular");
    origin.AddData(std::vector<float>{0, 0, 1});
    origin.AddData(std::vector<float>{0, 1, 0});
    origin.AddData(std::vector<float>{0, 0, 1});
    origin.Fit();
    std::vector<std::pair<int, float> > result;
    int ef_search = 3*10;
    origin.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);
    EXPECT_EQ(3, result.size());
    n2::Hnsw copy(3, "angular");
    copy = std::move(origin);
    std::vector<std::pair<int, float> > result2;
    copy.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result2);
    EXPECT_EQ(3, result2.size());
}

TEST_F(CppApiTest, MoveOperatorTest) {
    n2::Hnsw* origin = new n2::Hnsw(3, "angular");
    origin->AddData(std::vector<float>{0, 0, 1});
    origin->AddData(std::vector<float>{0, 1, 0});
    origin->AddData(std::vector<float>{0, 0, 1});
    origin->Fit();
    std::vector<std::pair<int, float> > result;
    int ef_search = 3*10;
    origin->SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);
    EXPECT_EQ(3, result.size());
    n2::Hnsw* copy = new n2::Hnsw(std::move(*origin));
    std::vector<std::pair<int, float> > result2;
    copy->SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result2);
    EXPECT_EQ(3, result2.size());
    delete copy;
    delete origin;
}

TEST_F(CppApiTest, LoadModelTest) {
    n2::Hnsw* origin = new n2::Hnsw();
    origin->LoadModel("../model/test.n2", false);
    std::vector<std::pair<int, float> > result;
    int ef_search = 3*10;
    origin->SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);
    EXPECT_EQ(3, result.size());
    delete origin;
}

n2::Hnsw load_n2_model(std::string path, bool use_mmap) {
    auto n2 = n2::Hnsw();
    n2.LoadModel(path, use_mmap);
    return std::move(n2);
}

TEST_F(CppApiTest, LoadModelMoveTestWithMmap) {
    auto origin = std::move(load_n2_model("../model/test.n2", true));
    n2::Hnsw other = std::move(origin);
    int ef_search = 3*10;
  
    for(size_t i = 0; i < 100000; i++ ) { 
        std::vector<std::pair<int, float> > result; 
        other.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);
        EXPECT_EQ(3, result.size());
    }
}

TEST_F(CppApiTest, LoadModelMoveTestWithInMemory) {
    auto origin = std::move(load_n2_model("../model/test.n2", false));
    n2::Hnsw other = std::move(origin);
    int ef_search = 3*10;

    for(size_t i = 0; i < 100000; i++ ) {
        std::vector<std::pair<int, float> > result;
        other.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);
        EXPECT_EQ(3, result.size());
    }
}

TEST_F(CppApiTest, LoadModelMoveTestAndUnloadWithInMmap) {
    auto origin = std::move(load_n2_model("../model/test.n2", true));
    n2::Hnsw other = n2::Hnsw();
    other  = std::move(origin);
    int ef_search = 3*10;

    std::vector<std::pair<int, float> > result;
    other.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);
    EXPECT_EQ(3, result.size());
    origin.UnloadModel();
}

TEST_F(CppApiTest, LoadModelMoveTestAndUnloadWithInMemory) {
    auto origin = std::move(load_n2_model("../model/test.n2", false));
    n2::Hnsw other = std::move(origin);
    int ef_search = 3*10;

    std::vector<std::pair<int, float> > result;
    other.SearchByVector(std::vector<float>{3, 2, 1}, 3, ef_search, result);
    EXPECT_EQ(3, result.size());
    origin.UnloadModel();
}

TEST_F(CppApiTest, UnloadMmapModelTest) {
    n2::Hnsw* origin = new n2::Hnsw(3);
    origin->LoadModel("../model/test.n2", true);
    origin->UnloadModel();
    delete origin;
}

TEST_F(CppApiTest, UnloadModelTest) {
    n2::Hnsw* origin = new n2::Hnsw(3);
    origin->LoadModel("../model/test.n2", false);
    origin->UnloadModel();
    delete origin;
}

TEST_F(CppApiTest, L2DistanceTest) {
    n2::BaseDistance* metric = new n2::L2Distance();
    float PORTABLE_ALIGN32 TmpRes[8];
    float vec1[] = {0.0, 0.0, 0.0};
    float vec2[] = {1,0, 0.0, 1.0};
    float vec3[] = {0,0, 0.75, 1.0};
    float res1 = metric->Evaluate(vec1, vec2, 3, TmpRes);
    EXPECT_FLOAT_EQ(1, res1);
    float res2 = metric->Evaluate(vec2, vec3, 3, TmpRes);
    EXPECT_FLOAT_EQ(1.5625, res2);
    delete metric;
}

TEST_F(CppApiTest, AngularDistanceTest) {
    n2::BaseDistance* metric = new n2::AngularDistance();
    float PORTABLE_ALIGN32 TmpRes[8];
    float vec1[] = {0.1, 0.2, 0.3, 0.4};
    float vec2[] = {0.5, 0.6, 0.7, 0.8};
    float res1 = metric->Evaluate(vec1, vec2, 4, TmpRes);
    delete metric;
    EXPECT_FLOAT_EQ(0.3, res1);
}

TEST_F(CppApiTest, MinHeapTest) {
    n2::MinHeap<int, float>* minheap = new n2::MinHeap<int, float>();
    minheap->push(3, 3.5);
    minheap->push(2, 7.5);
    minheap->push(1, 2.4);
    EXPECT_EQ(minheap->size(), 3);
    auto item = minheap->top(); minheap->pop();
    EXPECT_FLOAT_EQ(item.data, 2.4);
    minheap->pop();
    minheap->pop();
    EXPECT_EQ(minheap->size(), 0);
    delete minheap;
}

TEST_F(CppApiTest, DimensionErrorTest) {
    n2::Hnsw* origin = new n2::Hnsw(0, "angular");
    EXPECT_THROW(origin->AddData(std::vector<float>{0, 0, 1}),
    std::runtime_error);
    EXPECT_NO_THROW(origin->AddData(std::vector<float>{}));
    delete origin;
}

TEST_F(CppApiTest, LoadUnknownModelTest) {
    n2::Hnsw* origin = new n2::Hnsw(0);
    origin->LoadModel("../model/test.n2", false);
    EXPECT_THROW(origin->AddData(std::vector<float>{0, 0, 1}),
    std::runtime_error);
    EXPECT_THROW(origin->AddData(std::vector<float>{}),
    std::runtime_error);
    delete origin;
}

File: tests/golang_test/n2_test.go
// Copyright 2017 Kakao Corp. <http://www.kakaocorp.com>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package n2_test

import (
	"n2"
	"math/rand"
	"os"
	"testing"
)

func FloatArrayEquals(a []float32, b []float32) bool {
	if len(a) != len(b) {
		return false
	}
	for i, v := range a {
		if v != b[i] {
			return false
		}
	}
	return true
}

func IntArrayEquals(a []int, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i, v := range a {
		if v != b[i] {
			return false
		}
	}
	return true
}

func TestSearchByVectorL2(t *testing.T) {
	f := 3
	index := n2.NewHnswIndex(f, "L2")
	index.AddData([]float32{0, 0, 1})
	index.AddData([]float32{0, 1, 0})
	index.AddData([]float32{1, 0, 0})
	index.Build(5, 10, 4, 10, 3.5, "heuristic", "skip")
	index.SaveModel("test.n2")

	var result []int
	var distance []float32
	index.SearchByVector([]float32{1, 2, 3}, 3, -1, &result, &distance)
	if len(result) != 3 {
		t.Errorf("Result should be 3 not %d", len(result))
	}
	expected := []int{0, 1, 2}
	expected_distances := []float32{9.0, 11.0, 13.0}
	if IntArrayEquals(expected, result) != true {
		t.Errorf("Expected %v but got %v", expected, result)
	}

	if FloatArrayEquals(expected_distances, distance) != true {
		t.Errorf("Expected %v but got %v", expected_distances, distance)
	}

	n2.DeleteHnswIndex(index)
	index = n2.NewHnswIndex(f, "L2")
	if ret := index.LoadModel("test.n2"); ret == false {
		t.Errorf("Failed to load file")
	}

	n2.DeleteHnswIndex(index)

	index = n2.NewHnswIndex(f, "L2")
	if ret := index.LoadModel("test.n2", false); ret == false {
		t.Errorf("Failed to load file")
	}

	n2.DeleteHnswIndex(index)
	os.Remove("test.n2")
}

func TestSearchByVectorAngular(t *testing.T) {
	f := 3
	index := n2.NewHnswIndex(f, "angular")
	index.AddData([]float32{0, 0, 1})
	index.AddData([]float32{0, 1, 0})
	index.AddData([]float32{1, 0, 0})
	index.Build(5, 10, 4, 10, 3.5, "heuristic", "skip")

	var result []int

	expected1 := []int{0, 1, 2}
	expected2 := []int{0, 1, 2}
	expected3 := []int{0, 1, 2}
	index.SearchByVector([]float32{3, 2, 1}, 3, -1, &result)
	if len(result) != 3 {
		t.Errorf("The length of the result should be 3 not %d: %v", len(result), result)
	}

	if IntArrayEquals(expected1, result) != true {
		t.Errorf("Expected %v but got %v", expected1, result)
	}

	index.SearchByVector([]float32{1, 2, 3}, 3, -1, &result)
	if len(result) != 3 {
		t.Errorf("The length of the result should be 3 not %d: %v", len(result), result)
	}

	if IntArrayEquals(expected2, result) != true {
		t.Errorf("Expected %v but got %v", expected2, result)
	}

	index.SearchByVector([]float32{2, 0, 1}, 3, -1, &result)
	if len(result) != 3 {
		t.Errorf("Result should be 3 not %d", len(result))
	}

	if IntArrayEquals(expected3, result) != true {
		t.Errorf("Expected %v but got %v", expected3, result)
	}

	n2.DeleteHnswIndex(index)
}

func TestSearchByIdAngular(t *testing.T) {
	f := 3
	index := n2.NewHnswIndex(f, "angular")
	index.AddData([]float32{2, 1, 0})
	index.AddData([]float32{1, 2, 0})
	index.AddData([]float32{0, 0, 1})
	index.Build(5, 10, 4, 1, 3.5, "heuristic", "skip")

	var result []int

	expected1 := []int{0, 1, 2}
	expected2 := []int{1, 0, 2}
	index.SearchById(0, 3, -1, &result)
	if len(result) != 3 {
		t.Errorf("Result should be 3 not %d", len(result))
	}

	if IntArrayEquals(expected1, result) != true {
		t.Errorf("Expected %v but got %v", expected1, result)
	}

	index.SearchById(1, 3, -1, &result)
	if len(result) != 3 {
		t.Errorf("A length of the result should be 3 not %d", len(result))
	}

	if IntArrayEquals(expected2, result) != true {
		t.Errorf("Expected %v but got %v", expected2, result)
	}

	n2.DeleteHnswIndex(index)
}
func TestLargeL2(t *testing.T) {
	f := 10
	index := n2.NewHnswIndex(f, "L2")
	for j := 0; j < 10000; j += 2 {
		p := make([]float32, 0, 10)
		for i := 0; i < 10; i++ {
			p = append(p, rand.Float32())
		}
		x := make([]float32, 0, 10)
		for i := 0; i < 10; i++ {
			x = append(x, 1+p[i]+rand.Float32()*1e-2)
		}
		y := make([]float32, 0, 10)
		for i := 0; i < 10; i++ {
			y = append(y, 1+p[i]+rand.Float32()*1e-2)
		}
		index.AddData(x)
		index.AddData(y)
	}

	index.Build(5, 10, 300, 10, 3.5, "heuristic", "skip")

	for j := 0; j < 10000; j += 2 {
		var result []int
		index.SearchById(j, 2, -1, &result)
		if len(result) != 2 {
			t.Errorf("length of the result should be 2 not %d", len(result))
		}

		expected1 := []int{j, j + 1}
		if IntArrayEquals(expected1, result) != true {
			t.Errorf("Expected %v but got %v", expected1, result)
		}

		index.SearchById(j+1, 2, -1, &result)
		if len(result) != 2 {
			t.Errorf("length of the result should be 2 not %d", len(result))
		}
		expected2 := []int{j + 1, j}
		if IntArrayEquals(expected2, result) != true {
			t.Errorf("Expected %v but got %v", expected2, result)
		}
	}
	n2.DeleteHnswIndex(index)
}
